File: packages/python/bowline-client/src/bowline_client/transport.py
from __future__ import annotations

import json
from collections.abc import AsyncIterator, Callable, Iterator, Mapping
from typing import Any, BinaryIO, TypeVar

import httpx
from pydantic import BaseModel, TypeAdapter

from .errors import BowlineError, Code
from .sse import Event, aevents, events
from .types import CallOptions, Method

T = TypeVar("T")

HeadersSource = Callable[[], Mapping[str, str]]


def _encode(input: BaseModel) -> dict[str, Any]:
    return input.model_dump(mode="json", exclude_unset=True, by_alias=True)


def _decode(output: type[T], body: bytes) -> T:
    if not body.strip():
        body = b"{}"
    return TypeAdapter(output).validate_json(body)


def _headers(
    static: HeadersSource | None, options: CallOptions | None, accept: str
) -> dict[str, str]:
    merged: dict[str, str] = {}
    if static is not None:
        merged.update(static())
    if options is not None and options.headers:
        merged.update(options.headers)
    merged["accept"] = accept
    return merged


def _request(
    base: str, path: str, method: Method, input: BaseModel
) -> tuple[str, str, dict[str, str] | None, bytes | None]:
    url = f"{base}/{path}"
    payload = json.dumps(_encode(input), separators=(",", ":"))
    if method is Method.GET:
        return "GET", url, {"input": payload}, None
    return "POST", url, None, payload.encode()


def _timeout(options: CallOptions | None) -> httpx.Timeout | httpx._client.UseClientDefault:
    if options is not None and options.timeout is not None:
        return httpx.Timeout(options.timeout)
    return httpx.USE_CLIENT_DEFAULT


def _failure(path: str, exc: Exception) -> BowlineError:
    if isinstance(exc, httpx.TimeoutException):
        return BowlineError(Code.DEADLINE_EXCEEDED, f"timeout calling {path}", 0)
    return BowlineError(Code.UNAVAILABLE, f"network error calling {path}: {exc}", 0)


def _message(event: Event, output: type[T]) -> T | None:
    if event.name == "message":
        return TypeAdapter(output).validate_json(event.data)
    if event.name == "error":
        raise BowlineError.from_envelope(0, event.data)
    return None


def _multipart(input: BaseModel, file: BinaryIO, filename: str) -> list[tuple[str, Any]]:
    payload = json.dumps(_encode(input), separators=(",", ":")).encode()
    return [
        ("input", (None, payload, "application/json")),
        ("file", (filename, file, "application/octet-stream")),
    ]


class Transport:
    def __init__(
        self,
        base_url: str,
        *,
        client: httpx.AsyncClient | None = None,
        headers: HeadersSource | None = None,
    ) -> None:
        self._base = base_url.rstrip("/")
        self._client = client or httpx.AsyncClient()
        self._headers = headers

    async def call(
        self,
        path: str,
        method: Method,
        input: BaseModel,
        output: type[T],
        options: CallOptions | None = None,
    ) -> T:
        verb, url, params, content = _request(self._base, path, method, input)
        headers = _headers(self._headers, options, "application/json")
        if content is not None:
            headers["content-type"] = "application/json"
        try:
            response = await self._client.request(
                verb,
                url,
                params=params,
                content=content,
                headers=headers,
                timeout=_timeout(options),
            )
        except httpx.HTTPError as exc:
            raise _failure(path, exc) from exc
        if response.status_code >= 400:
            raise BowlineError.from_envelope(response.status_code, response.content)
        return _decode(output, response.content)

    async def subscribe(
        self,
        path: str,
        input: BaseModel,
        output: type[T],
        options: CallOptions | None = None,
        method: Method = Method.GET,
    ) -> AsyncIterator[T]:
        verb, url, params, content = _request(self._base, path, method, input)
        headers = _headers(self._headers, options, "text/event-stream")
        if content is not None:
            headers["content-type"] = "application/json"
        try:
            async with self._client.stream(
                verb,
                url,
                params=params,
                content=content,
                headers=headers,
                timeout=_timeout(options),
            ) as response:
                if response.status_code >= 400:
                    body = await response.aread()
                    raise BowlineError.from_envelope(response.status_code, body)
                async for event in aevents(response.aiter_lines()):
                    if event.name == "done":
                        return
                    value = _message(event, output)
                    if value is not None:
                        yield value
        except httpx.HTTPError as exc:
            raise _failure(path, exc) from exc

    async def upload(
        self,
        path: str,
        input: BaseModel,
        file: BinaryIO,
        filename: str,
        output: type[T],
        options: CallOptions | None = None,
    ) -> T:
        headers = _headers(self._headers, options, "application/json")
        try:
            response = await self._client.post(
                f"{self._base}/{path}",
                files=_multipart(input, file, filename),
                headers=headers,
                timeout=_timeout(options),
            )
        except httpx.HTTPError as exc:
            raise _failure(path, exc) from exc
        if response.status_code >= 400:
            raise BowlineError.from_envelope(response.status_code, response.content)
        return _decode(output, response.content)

    async def aclose(self) -> None:
        await self._client.aclose()


class SyncTransport:
    def __init__(
        self,
        base_url: str,
        *,
        client: httpx.Client | None = None,
        headers: HeadersSource | None = None,
    ) -> None:
        self._base = base_url.rstrip("/")
        self._client = client or httpx.Client()
        self._headers = headers

    def call(
        self,
        path: str,
        method: Method,
        input: BaseModel,
        output: type[T],
        options: CallOptions | None = None,
    ) -> T:
        verb, url, params, content = _request(self._base, path, method, input)
        headers = _headers(self._headers, options, "application/json")
        if content is not None:
            headers["content-type"] = "application/json"
        try:
            response = self._client.request(
                verb,
                url,
                params=params,
                content=content,
                headers=headers,
                timeout=_timeout(options),
            )
        except httpx.HTTPError as exc:
            raise _failure(path, exc) from exc
        if response.status_code >= 400:
            raise BowlineError.from_envelope(response.status_code, response.content)
        return _decode(output, response.content)

    def subscribe(
        self,
        path: str,
        input: BaseModel,
        output: type[T],
        options: CallOptions | None = None,
        method: Method = Method.GET,
    ) -> Iterator[T]:
        verb, url, params, content = _request(self._base, path, method, input)
        headers = _headers(self._headers, options, "text/event-stream")
        if content is not None:
            headers["content-type"] = "application/json"
        try:
            with self._client.stream(
                verb,
                url,
                params=params,
                content=content,
                headers=headers,
                timeout=_timeout(options),
            ) as response:
                if response.status_code >= 400:
                    body = response.read()
                    raise BowlineError.from_envelope(response.status_code, body)
                for event in events(response.iter_lines()):
                    if event.name == "done":
                        return
                    value = _message(event, output)
                    if value is not None:
                        yield value
        except httpx.HTTPError as exc:
            raise _failure(path, exc) from exc

    def upload(
        self,
        path: str,
        input: BaseModel,
        file: BinaryIO,
        filename: str,
        output: type[T],
        options: CallOptions | None = None,
    ) -> T:
        headers = _headers(self._headers, options, "application/json")
        try:
            response = self._client.post(
                f"{self._base}/{path}",
                files=_multipart(input, file, filename),
                headers=headers,
                timeout=_timeout(options),
            )
        except httpx.HTTPError as exc:
            raise _failure(path, exc) from exc
        if response.status_code >= 400:
            raise BowlineError.from_envelope(response.status_code, response.content)
        return _decode(output, response.content)

    def close(self) -> None:
        self._client.close()

File: packages/dart/bowline/lib/src/transport.dart
import 'dart:async';
import 'dart:convert';

import 'package:http/http.dart' as http;

import 'error.dart';
import 'sse.dart';

enum Method { get, post, put, patch, delete }

class CallOptions {
  const CallOptions({this.headers, this.timeout});

  final Map<String, String>? headers;
  final Duration? timeout;
}

class Transport {
  Transport(
      {required Uri baseUrl,
      http.Client? client,
      Map<String, String> Function()? headers})
      : _base = baseUrl.toString().replaceAll(RegExp(r'/+$'), ''),
        _client = client ?? http.Client(),
        _headers = headers;

  final String _base;
  final http.Client _client;
  final Map<String, String> Function()? _headers;

  Future<T> call<T>(
    String path,
    Method method,
    Map<String, Object?> input,
    T Function(Object?) decode, {
    CallOptions? options,
    bool rest = false,
  }) async {
    final request =
        _request(path, method, input, options, 'application/json', rest);
    final response = await _send(request, path, options);
    final body = await _read(response, path, options);
    if (response.statusCode >= 400) {
      throw _failure(response.statusCode, body, path);
    }
    return decode(body.isEmpty ? null : jsonDecode(body));
  }

  Stream<T> subscribe<T>(
    String path,
    Map<String, Object?> input,
    T Function(Object?) decode, {
    CallOptions? options,
  }) {
    final controller = StreamController<T>();
    StreamSubscription<ServerEvent>? events;
    var finished = false;
    void finish([Object? error]) {
      if (finished) {
        return;
      }
      finished = true;
      if (error != null) {
        controller.addError(error);
      }
      unawaited(events?.cancel());
      unawaited(controller.close());
    }

    controller.onListen = () async {
      try {
        final request = _request(
            path, Method.get, input, options, 'text/event-stream', false);
        final response = await _send(request, path, options);
        if (response.statusCode >= 400) {
          finish(_failure(
              response.statusCode, await _read(response, path, options), path));
          return;
        }
        events = parseServerSentEvents(response.stream).listen(
          (event) {
            switch (event.name) {
              case 'message':
                try {
                  controller.add(decode(jsonDecode(event.data)));
                } catch (e) {
                  finish(e);
                }
              case 'error':
                finish(_failure(0, event.data, path));
              case 'done':
                finish();
            }
          },
          onError: finish,
          onDone: finish,
        );
      } catch (e) {
        finish(e);
      }
    };
    controller.onCancel = () {
      finished = true;
      unawaited(events?.cancel());
    };
    return controller.stream;
  }

  Future<T> upload<T>(
    String path,
    Map<String, Object?> input,
    Stream<List<int>> file,
    String filename,
    T Function(Object?) decode, {
    CallOptions? options,
  }) async {
    final request = http.MultipartRequest('POST', Uri.parse('$_base/$path'));
    request.headers.addAll(_mergedHeaders(options, 'application/json'));
    request.files.add(http.MultipartFile.fromString('input', jsonEncode(input),
        contentType: _json));
    final bytes = await http.ByteStream(file).toBytes();
    request.files
        .add(http.MultipartFile.fromBytes('file', bytes, filename: filename));
    final response = await _send(request, path, options);
    final body = await _read(response, path, options);
    if (response.statusCode >= 400) {
      throw _failure(response.statusCode, body, path);
    }
    return decode(body.isEmpty ? null : jsonDecode(body));
  }

  static final _json = http.MediaType('application', 'json');

  Map<String, String> _mergedHeaders(CallOptions? options, String accept) {
    final headers = <String, String>{
      ...?_headers?.call(),
      ...?options?.headers
    };
    headers['accept'] = accept;
    return headers;
  }

  http.Request _request(
    String path,
    Method method,
    Map<String, Object?> input,
    CallOptions? options,
    String accept,
    bool rest,
  ) {
    final verb = method.name.toUpperCase();
    final url = Uri.parse('$_base/$path');
    final http.Request request;
    if (_sendsBody(method)) {
      request = http.Request(verb, url)..body = jsonEncode(input);
    } else if (rest) {
      request = http.Request(verb, _withQuery(url, input));
    } else {
      request = http.Request(
          verb, url.replace(queryParameters: {'input': jsonEncode(input)}));
    }
    request.headers.addAll(_mergedHeaders(options, accept));
    if (_sendsBody(method)) {
      request.headers['content-type'] = 'application/json';
    }
    return request;
  }

  static bool _sendsBody(Method method) =>
      method != Method.get && method != Method.delete;

  static Uri _withQuery(Uri url, Map<String, Object?> input) {
    final params = <String, List<String>>{};
    input.forEach((key, value) {
      if (value == null) {
        return;
      }
      final values = value is Iterable
          ? [for (final item in value) item.toString()]
          : [value.toString()];
      if (values.isNotEmpty) {
        params[key] = values;
      }
    });
    return params.isEmpty ? url : url.replace(queryParameters: params);
  }

  Future<http.StreamedResponse> _send(
      http.BaseRequest request, String path, CallOptions? options) async {
    try {
      var pending = _client.send(request);
      final timeout = options?.timeout;
      if (timeout != null) {
        pending = pending.timeout(timeout);
      }
      return await pending;
    } on TimeoutException {
      throw BowlineException(Code.deadlineExceeded, 'calling $path timed out');
    } on http.ClientException catch (e) {
      throw BowlineException(Code.unavailable, 'calling $path: ${e.message}');
    } on BowlineException {
      rethrow;
    } catch (e) {
      throw BowlineException(Code.unavailable, 'calling $path: $e');
    }
  }

  Future<String> _read(
      http.StreamedResponse response, String path, CallOptions? options) async {
    try {
      var pending = response.stream.bytesToString();
      final timeout = options?.timeout;
      if (timeout != null) {
        pending = pending.timeout(timeout);
      }
      return await pending;
    } on TimeoutException {
      throw BowlineException(Code.deadlineExceeded, 'reading $path timed out');
    } on http.ClientException catch (e) {
      throw BowlineException(Code.unavailable, 'reading $path: ${e.message}');
    }
  }

  BowlineException _failure(int status, String body, String path) {
    Object? decoded;
    try {
      decoded = jsonDecode(body);
    } on FormatException {
      decoded = null;
    }
    return BowlineException.fromEnvelope(decoded, status) ??
        BowlineException(Code.unknown, 'HTTP $status from $path',
            status: status);
  }
}
